In [1]:
import pandas as pd
from datetime import datetime

# READ DATA

In [2]:
df_2021 = pd.read_excel('Historical_sales2021.xlsx')
df_2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 409368 entries, 0 to 409367
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    409368 non-null  int64         
 1   transaction_date  409368 non-null  datetime64[ns]
 2   transaction_time  409368 non-null  object        
 3   store_id          409368 non-null  int64         
 4   staff_id          409368 non-null  int64         
 5   customer_id       409368 non-null  int64         
 6   product_id        409368 non-null  int64         
 7   quantity_sold     409368 non-null  int64         
 8   unit_price        409368 non-null  float64       
 9   promo_item_yn     409368 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(6), object(2)
memory usage: 31.2+ MB


In [3]:
df_2022 = pd.read_csv('Historical_sales2022 .csv')
df_2022.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140445 entries, 0 to 140444
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   transaction_id    140445 non-null  int64  
 1   transaction_date  140445 non-null  object 
 2   transaction_time  140445 non-null  object 
 3   store_id          140445 non-null  int64  
 4   staff_id          140445 non-null  int64  
 5   customer_id       140445 non-null  int64  
 6   product_id        140445 non-null  int64  
 7   quantity_sold     140445 non-null  int64  
 8   unit_price        140445 non-null  float64
 9   promo_item_yn     140445 non-null  object 
dtypes: float64(1), int64(6), object(3)
memory usage: 10.7+ MB


In [4]:
df_2022.duplicated().sum()

408

In [5]:
df_2021.duplicated().sum()

1189

In [6]:
df_empl = pd.read_csv('Dim_employee.csv')
df_empl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55 entries, 0 to 54
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   staff_id    55 non-null     int64 
 1   first_name  55 non-null     object
 2   last_name   55 non-null     object
 3   position    55 non-null     object
 4   start_date  55 non-null     object
 5   end_date    55 non-null     object
dtypes: int64(1), object(5)
memory usage: 2.7+ KB


In [7]:
df_empl.head()

staff_id first_name last_name      position  start_date  end_date
0         1       Mark    Brewer           CFO    8/3/2001  1/1/3000
1         2       Jean    LeBean           CEO    8/3/2001  1/1/3000
2         3      Jamie     Toast  Head Roaster  10/24/2007  1/1/3000
3         4    Chelsea   Claudia       Roaster    7/3/2003  1/1/3000
4         5       Adam     Songs  Head Barista    4/2/2008  1/1/3000

# Preprocessing

## Drop_duplicate 

In [8]:
df_2021.drop_duplicates(inplace=True)
df_2021['sales'] = df_2021['quantity_sold'] * df_2021['unit_price']
df_2021.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 408179 entries, 0 to 409367
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    408179 non-null  int64         
 1   transaction_date  408179 non-null  datetime64[ns]
 2   transaction_time  408179 non-null  object        
 3   store_id          408179 non-null  int64         
 4   staff_id          408179 non-null  int64         
 5   customer_id       408179 non-null  int64         
 6   product_id        408179 non-null  int64         
 7   quantity_sold     408179 non-null  int64         
 8   unit_price        408179 non-null  float64       
 9   promo_item_yn     408179 non-null  object        
 10  sales             408179 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(6), object(2)
memory usage: 37.4+ MB


In [9]:
df_2022[(df_2022['transaction_id']==1) & (df_2022['staff_id']==45)]

transaction_id transaction_date transaction_time  store_id  staff_id  \
13204                1        1/14/2022          8:48:00         8        45   
28588                1        1/29/2022          8:48:00         8        45   
51164                1        2/21/2022          8:35:00         8        45   
51165                1        2/21/2022          8:35:00         8        45   
71005                1        3/13/2022          8:05:00         8        45   
72237                1        3/14/2022          8:48:00         8        45   
88045                1        3/29/2022          8:48:00         8        45   
111087               1        4/13/2022          8:05:00         8        45   
113084               1        4/14/2022          8:48:00         8        45   
125639               1        4/21/2022          8:35:00         8        45   
125640               1        4/21/2022          8:35:00         8        45   
139148               1        4/29/2022          8:48:00         8        45   

        customer_id  product_id  quantity_sold  unit_price promo_item_yn  
13204          8600          29              1         2.5             N  
28588          8600          87              2         2.1             Y  
51164          8600          48              2         2.5             N  
51165          8600          81              1        28.0             N  
71005          8294          57              2         3.1             N  
72237          8600          29              1         2.5             N  
88045          8600          87              2         2.1             Y  
111087         8294          57              2         3.1             N  
113084         8600          29              1         2.5             N  
125639         8600          48              2         2.5             N  
125640         8600          81              1        28.0             N  
139148         8600          87              2         2.1             Y

In [10]:
df_2022.drop_duplicates(inplace=True)
df_2022['sales'] = df_2022['quantity_sold'] * df_2022['unit_price']
df_2022.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 140037 entries, 0 to 140444
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   transaction_id    140037 non-null  int64  
 1   transaction_date  140037 non-null  object 
 2   transaction_time  140037 non-null  object 
 3   store_id          140037 non-null  int64  
 4   staff_id          140037 non-null  int64  
 5   customer_id       140037 non-null  int64  
 6   product_id        140037 non-null  int64  
 7   quantity_sold     140037 non-null  int64  
 8   unit_price        140037 non-null  float64
 9   promo_item_yn     140037 non-null  object 
 10  sales             140037 non-null  float64
dtypes: float64(2), int64(6), object(3)
memory usage: 12.8+ MB


In [11]:
df_2021.head()

transaction_id transaction_date transaction_time  store_id  staff_id  \
0             390       2021-01-01         07:01:20         5        12   
1            2465       2021-01-01         07:02:34         5        12   
2            1644       2021-01-01         07:03:30         5        12   
3            1034       2021-01-01         07:04:24         5        30   
4             799       2021-01-01         07:06:29         5        12   

   customer_id  product_id  quantity_sold  unit_price promo_item_yn  sales  
0         5658          22              2         2.0             N    4.0  
1         5163          30              2         3.0             N    6.0  
2         5247          22              2         2.0             N    4.0  
3         5333          37              1         3.0             N    3.0  
4         5923          46              1         2.5             N    2.5

# REPORT

### 2021

In [12]:
# filter sales in 2021 Q1
df_2021_Q1 = df_2021[(df_2021['transaction_date'] >= '2021-01-01') & (df_2021['transaction_date'] <= '2021-03-31')]
df_2021_Q1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87272 entries, 0 to 87522
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    87272 non-null  int64         
 1   transaction_date  87272 non-null  datetime64[ns]
 2   transaction_time  87272 non-null  object        
 3   store_id          87272 non-null  int64         
 4   staff_id          87272 non-null  int64         
 5   customer_id       87272 non-null  int64         
 6   product_id        87272 non-null  int64         
 7   quantity_sold     87272 non-null  int64         
 8   unit_price        87272 non-null  float64       
 9   promo_item_yn     87272 non-null  object        
 10  sales             87272 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(6), object(2)
memory usage: 8.0+ MB


In [13]:
#Pivot_table for 2021 Q1
df_pivot_2021 = df_2021_Q1.pivot_table(index='staff_id', values='sales', aggfunc='sum').sort_values(by='sales', ascending=False)
df_pivot_2021 = pd.merge(df_pivot_2021, df_empl[['first_name','last_name', 'staff_id']], on='staff_id', how='left')
df_pivot_2021.head()

staff_id     sales first_name last_name
0        12  49026.80   Britanni    Jorden
1        42  32542.14      Kylie   Candace
2        26  30545.80     Joelle  Christen
3        45  27895.14    Pandora   Neville
4        16  26020.36     Darren        Xu

COMMISSION OF EACH STAFF

In [14]:
# Calculate commission for each staff
df_pivot_2021['commission'] = df_pivot_2021['sales'] * 0.05
#rearrage columns
df_pivot_2021 = df_pivot_2021[['staff_id', 'first_name', 'last_name', 'sales', 'commission']]
df_pivot_2021.head()

staff_id first_name last_name     sales  commission
0        12   Britanni    Jorden  49026.80    2451.340
1        42      Kylie   Candace  32542.14    1627.107
2        26     Joelle  Christen  30545.80    1527.290
3        45    Pandora   Neville  27895.14    1394.757
4        16     Darren        Xu  26020.36    1301.018

=> Britanni	Jorden is the staff having highets commission in Q1 2021

In [37]:
# calculate percentage of sales for each staff
df_pivot_2021['sales_percentage'] = (round(df_pivot_2021['sales'] / sum(df_pivot_2021['sales'])*100,2)).astype(str) + '%'
df_pivot_2021

staff_id first_name last_name     sales  commission sales_percentage
0         12   Britanni    Jorden  49026.80   2451.3400           12.02%
1         42      Kylie   Candace  32542.14   1627.1070            7.98%
2         26     Joelle  Christen  30545.80   1527.2900            7.49%
3         45    Pandora   Neville  27895.14   1394.7570            6.84%
4         16     Darren        Xu  26020.36   1301.0180            6.38%
5         17      Quail   Octavia  25026.73   1251.3365            6.14%
6         20      Ronan     Magee  24443.56   1222.1780            5.99%
7         14      Damon     Sasha  22090.56   1104.5280            5.42%
8         30      Amela  Chadwick  20076.88   1003.8440            4.92%
9         44    Tamekah      Maya  19306.42    965.3210            4.73%
10        29      Orson  Benedict  17928.77    896.4385             4.4%
11        27    Ainsley    Evelyn  17432.83    871.6415            4.27%
12        43      Tatum    Laurel  17139.31    856.9655             4.2%
13        28     Joseph     Byron  15822.16    791.1080            3.88%
14        15   Remedios      Mari  15519.58    775.9790            3.81%
15         6      Karen     Cupps  13043.91    652.1955             3.2%
16        41     Adrian     Macon   9720.03    486.0015            2.38%
17         7     Kelsey   Cameron   5519.21    275.9605            1.35%
18        19      Peter    Paloma   4783.56    239.1780            1.17%
19        18    Ezekiel    Rashad   3899.95    194.9975            0.96%
20         9   Caldwell      Veda   3057.10    152.8550            0.75%
21         8   Hamilton       Emi   2450.25    122.5125             0.6%
22        25      Aline   Melanie   2190.39    109.5195            0.54%
23        10        Ima  Winifred   1742.60     87.1300            0.43%
24        13       Berk     Derek    646.15     32.3075            0.16%

=> Britanni Jordan's contribution to the total sales was 12.02%.

In [32]:
total_sales_2021_q1 = sum(df_pivot_2021['sales'])
total_sales_2021_q1

407870.19000000006

In [33]:
len(df_pivot_2021['sales'])

25

In [34]:
# Average sales per staff
avg_sales_2021 = total_sales_2021_q1 / len(df_pivot_2021['sales'])
avg_sales_2021

16314.807600000002

### 2022

In [19]:
df_2022.head()

transaction_id transaction_date transaction_time  store_id  staff_id  \
0             390         1/1/2022          7:01:00         5        12   
1            1228         1/1/2022          7:02:00         5        30   
2            1644         1/1/2022          7:03:00         5        12   
3            1338         1/1/2022          7:06:00         5        30   
4            2257         1/1/2022          7:06:00         5        30   

   customer_id  product_id  quantity_sold  unit_price promo_item_yn  sales  
0         5658          22              2         2.0             N    4.0  
1         5197          22              1         2.0             N    2.0  
2         5247          22              2         2.0             N    4.0  
3         5465          32              2         3.0             N    6.0  
4         5627          22              1         2.0             N    2.0

In [20]:
#convert transaction_date to datetime
df_2022['transaction_date'] = pd.to_datetime(df_2022['transaction_date'])
df_2022.head()

transaction_id transaction_date transaction_time  store_id  staff_id  \
0             390       2022-01-01          7:01:00         5        12   
1            1228       2022-01-01          7:02:00         5        30   
2            1644       2022-01-01          7:03:00         5        12   
3            1338       2022-01-01          7:06:00         5        30   
4            2257       2022-01-01          7:06:00         5        30   

   customer_id  product_id  quantity_sold  unit_price promo_item_yn  sales  
0         5658          22              2         2.0             N    4.0  
1         5197          22              1         2.0             N    2.0  
2         5247          22              2         2.0             N    4.0  
3         5465          32              2         3.0             N    6.0  
4         5627          22              1         2.0             N    2.0

In [21]:
# filter sales in 2022 Q1
df_2022_Q1 = df_2022[(df_2022['transaction_date'] >= '2022-01-01') & (df_2022['transaction_date'] <= '2022-03-31')]

In [22]:
df_2022_Q1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 90464 entries, 0 to 90732
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    90464 non-null  int64         
 1   transaction_date  90464 non-null  datetime64[ns]
 2   transaction_time  90464 non-null  object        
 3   store_id          90464 non-null  int64         
 4   staff_id          90464 non-null  int64         
 5   customer_id       90464 non-null  int64         
 6   product_id        90464 non-null  int64         
 7   quantity_sold     90464 non-null  int64         
 8   unit_price        90464 non-null  float64       
 9   promo_item_yn     90464 non-null  object        
 10  sales             90464 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(6), object(2)
memory usage: 8.3+ MB


In [23]:
#Pivot_table for 2022 Q1
df_pivot_2022 = df_2022_Q1.pivot_table(index='staff_id', values='sales', aggfunc='sum').sort_values(by='sales', ascending=False)
df_pivot_2022 = pd.merge(df_pivot_2022, df_empl[['first_name','last_name', 'staff_id']], on='staff_id', how='left')
df_pivot_2022.head()

staff_id     sales first_name last_name
0        12  50137.29   Britanni    Jorden
1        42  34270.84      Kylie   Candace
2        26  30728.14     Joelle  Christen
3        45  29164.18    Pandora   Neville
4        16  26754.85     Darren        Xu

In [24]:
# Calculate commission for each staff
df_pivot_2022['commission'] = df_pivot_2022['sales'] * 0.05
#rearrage columns
df_pivot_2022 = df_pivot_2022[['staff_id', 'first_name', 'last_name', 'sales', 'commission']]
df_pivot_2022.head()

staff_id first_name last_name     sales  commission
0        12   Britanni    Jorden  50137.29   2506.8645
1        42      Kylie   Candace  34270.84   1713.5420
2        26     Joelle  Christen  30728.14   1536.4070
3        45    Pandora   Neville  29164.18   1458.2090
4        16     Darren        Xu  26754.85   1337.7425

In [39]:
# calculate percentage of sales for each staff
df_pivot_2022['sales_percentage'] = (round(df_pivot_2022['sales'] / sum(df_pivot_2022['sales'])*100,2)).astype(str) + '%'
df_pivot_2022

staff_id first_name last_name     sales  commission sales_percentage
0         12   Britanni    Jorden  50137.29   2506.8645            11.8%
1         42      Kylie   Candace  34270.84   1713.5420            8.07%
2         26     Joelle  Christen  30728.14   1536.4070            7.23%
3         45    Pandora   Neville  29164.18   1458.2090            6.87%
4         16     Darren        Xu  26754.85   1337.7425             6.3%
5         17      Quail   Octavia  26702.52   1335.1260            6.29%
6         20      Ronan     Magee  25544.31   1277.2155            6.01%
7         14      Damon     Sasha  23053.27   1152.6635            5.43%
8         30      Amela  Chadwick  20643.07   1032.1535            4.86%
9         44    Tamekah      Maya  20072.75   1003.6375            4.73%
10        29      Orson  Benedict  18839.62    941.9810            4.44%
11        27    Ainsley    Evelyn  18544.72    927.2360            4.37%
12        43      Tatum    Laurel  17890.46    894.5230            4.21%
13        15   Remedios      Mari  16260.01    813.0005            3.83%
14        28     Joseph     Byron  16146.51    807.3255             3.8%
15         6      Karen     Cupps  14689.53    734.4765            3.46%
16        41     Adrian     Macon  10099.31    504.9655            2.38%
17         7     Kelsey   Cameron   6051.75    302.5875            1.42%
18        19      Peter    Paloma   4917.71    245.8855            1.16%
19        18    Ezekiel    Rashad   4237.45    211.8725             1.0%
20         9   Caldwell      Veda   3080.80    154.0400            0.73%
21         8   Hamilton       Emi   2587.60    129.3800            0.61%
22        25      Aline   Melanie   1850.18     92.5090            0.44%
23        10        Ima  Winifred   1794.60     89.7300            0.42%
24        13       Berk     Derek    695.55     34.7775            0.16%

=> Britanni Jordan's contribution to the total sales was 11.80%.

In [26]:
df_pivot_2022

staff_id first_name last_name     sales  commission  sales_percentage
0         12   Britanni    Jorden  50137.29   2506.8645         11.803758
1         42      Kylie   Candace  34270.84   1713.5420          8.068340
2         26     Joelle  Christen  30728.14   1536.4070          7.234287
3         45    Pandora   Neville  29164.18   1458.2090          6.866085
4         16     Darren        Xu  26754.85   1337.7425          6.298860
5         17      Quail   Octavia  26702.52   1335.1260          6.286540
6         20      Ronan     Magee  25544.31   1277.2155          6.013864
7         14      Damon     Sasha  23053.27   1152.6635          5.427402
8         30      Amela  Chadwick  20643.07   1032.1535          4.859971
9         44    Tamekah      Maya  20072.75   1003.6375          4.725702
10        29      Orson  Benedict  18839.62    941.9810          4.435388
11        27    Ainsley    Evelyn  18544.72    927.2360          4.365960
12        43      Tatum    Laurel  17890.46    894.5230          4.211928
13        15   Remedios      Mari  16260.01    813.0005          3.828073
14        28     Joseph     Byron  16146.51    807.3255          3.801352
15         6      Karen     Cupps  14689.53    734.4765          3.458337
16        41     Adrian     Macon  10099.31    504.9655          2.377668
17         7     Kelsey   Cameron   6051.75    302.5875          1.424756
18        19      Peter    Paloma   4917.71    245.8855          1.157770
19        18    Ezekiel    Rashad   4237.45    211.8725          0.997617
20         9   Caldwell      Veda   3080.80    154.0400          0.725309
21         8   Hamilton       Emi   2587.60    129.3800          0.609195
22        25      Aline   Melanie   1850.18     92.5090          0.435586
23        10        Ima  Winifred   1794.60     89.7300          0.422500
24        13       Berk     Derek    695.55     34.7775          0.163752

In [27]:
total_sales_2022 = sum(df_pivot_2022['sales'])
total_sales_2022

424757.01999999996

In [28]:
#calculate average sales per staff
avg_sales_2022 = total_sales_2022 / len(df_pivot_2022['sales'])
avg_sales_2022

16990.280799999997

In [29]:
len(df_pivot_2022['sales'])

25

In [41]:
increase_sales = (total_sales_2022 - total_sales_2021_q1) / total_sales_2021_q1 *100
increase_sales

4.140246189602603